File: Chapter12/Swift/ArraySortingAlgorithms.swift
//
//  ArraySortingAlgorithms.swift
//  EverydayDataStructures-Swift
//
//  Created by William Smith on 10/26/16.
//  Copyright © 2016 Websmiths, LLC. All rights reserved.
//

import Foundation

public class ArraySortingAlgorithms
{
    open func swap( x: inout Int, y: inout Int)
    {
        let t: Int = x
        x = y
        y = t
    }
    
    //Insertion Sort
    open func insertionSort( values: inout [Int])
    {
        if (values.count <= 1)
        {
            return
        }
        
        var j, value: Int
        for i in 1..<values.count
        {
            value = values[i]
            j = i - 1

            while (j >= 0 && values[j] > value)
            {
                values[j + 1] = values[j]
                j = j - 1
            }
            values[j + 1] = value
        }
    }
    
    //Selection Sort
    open func selectionSort( values: inout [Int])
    {
        if (values.count <= 1)
        {
            return
        }
        
        var minIndex: Int
        for i in 0..<values.count
        {
            minIndex = i
            for j in i+1..<values.count
            {
                if (values[j] < values[minIndex])
                {
                    minIndex = j
                }
            }
            
            swap(x: &values[minIndex], y: &values[i])
        }
    }
    
    //Quick Sort
    open func quickSort( values: inout [Int], low: Int, high: Int)
    {
        if (low < high)
        {
            let index: Int = partition( values: &values, low: low, high: high)
            
            quickSort( values: &values, low: low, high: index - 1)
            quickSort( values: &values, low: index + 1, high: high)
        }
    }

    func partition( values: inout [Int], low: Int, high: Int) -> Int
    {
        let pivot: Int = values[high]
        var i: Int = (low - 1)
        var j: Int = low
        
        while j <= (high - 1)
        {
            if (values[j] <= pivot)
            {
                i += 1
                swap(x: &values[i], y: &values[j])
            }
            
            j += 1
        }
        
        i += 1
        swap(x: &values[i], y: &values[high])
        
        return i;
    }
    
    
    //Merge Sort
    open func mergeSort( values: inout [Int], left: Int, right: Int)
    {
        if (left == right)
        {
            return
        }
        
        if (left < right)
        {
            let middle: Int = (left + right) / 2
            
            mergeSort(values: &values, left: left, right: middle)
            mergeSort(values: &values, left: middle + 1, right: right)
            
            var temp = values
            
            var index1: Int = left
            var index2: Int = middle + 1
            for n in left...right
            {
                if (index1 == middle + 1)
                {
                    values[n] = temp[index2]
                    index2 += 1
                }
                else if (index2 > right)
                {
                    values[n] = temp[index1]
                    index1 += 1
                }
                else if (temp[index1] < temp[index2])
                {
                    values[n] = temp[index1]
                    index1 += 1
                }
                else
                {
                    values[n] = temp[index2]
                    index2 += 1
                }
            }
        }
    }
    
    //BubbleSort
    open func bubbleSort( values: inout [Int])
    {
        if (values.count <= 1)
        {
            return
        }
        
        var swapped: Bool
        var i: Int = 0
        while i < (values.count - 1)
        {
            swapped = false
            var j: Int = values.count - 1
            while j > i
            {
                if (values[j] < values[j - 1])
                {
                    swap(x: &values[j], y: &values[j - 1])
                    swapped = true
                }
                j -= 1
            }
        
            if (swapped == false)
            {
                break
            }
            i += 1
        }
    }
    
    //BucketSort
    open func bucketSort( values: inout [Int], maxVal: Int)
    {
        var bucket = [Int]()
        let num: Int = values.count
        let bucketNum: Int = maxVal + 1
        
        for i in 0..<bucketNum
        {
            bucket.insert(0, at: i)
        }
        
        for i in 0..<num
        {
            bucket[values[i]] = bucket[values[i]] + 1
        }
        
        var pos: Int = 0
        for i in 0..<bucketNum
        {
            for _ in 0..<bucket[i]
            {
                values[pos] = i
                pos += 1
            }
        }
    }
    
    //CountSort
    open func countingSort( values: inout [Int])
    {
        let n: Int = values.count
        var min: Int = values.min()!
        var max: Int = values.max()!
        
        for i in 1..<n
        {
            let iVal: Int = values[i]
            if (iVal < min)
            {
                min = iVal
            }
            else if (iVal > max)
            {
                max = iVal
            }
        }
        
        var count = [Int]()
        for i in 0..<max
        {
            count.insert(0, at: i)
        }
        
        for i in 0..<n
        {
            count[values[i] - min] = count[values[i] - min] + 1
        }
        
        var z: Int = 0
        var i: Int = min
        while i <= max
        {
            while (count[i - min] > 0)
            {
                values[z] = i
                z += 1
                count[i - min] = count[i - min] - 1
            }
            i += 1
        }
    }
    
    
    //RadixSort
    open func radixSort( values: inout [Int])
    {
        let n: Int = values.count
        var max: Int = values[0]
        
        for i in 1..<n
        {
            if (values[i] > max)
            {
                max = values[i];
            }
        }
        
        var exp: Int = 1
        while max/exp > 0 {
            var output: [Int] = [Int]()
            for _ in 0..<n
            {
                output.append(0)
            }
            
            var count: [Int] = [Int]()
            for _ in 0..<10
            {
                count.append(0)
            }
            
            for i in 0..<n
            {
                let index: Int = (values[i] / exp) % 10
                count[index] = count[index] + 1
            }
            
            for i in 1..<10
            {
                count[i] = count[i] + count[i - 1]
            }
            
            var i: Int = n - 1
            while i >= 0
            {
                let index: Int = (values[i] / exp) % 10
                output[count[index] - 1] = values[i]
                count[index] = count[index] - 1
                i -= 1
            }
            
            for i in 0..<n
            {
                values[i] = output[i];
            }
            exp *= 10
        }
    }
    
}

File: Chapter12/Java/ArraySortingAlgorithms.java
package main.java;

/**
 * Created by William Smith on 10/26/16.
 */
public class ArraySortingAlgorithms
{
    //InsertionSort
    public void insertionSort(int[] values)
    {
        if (values.length <= 1)
            return;

        int j, value;
        for (int i = 1; i < values.length; i++)
        {
            value = values[i];
            j = i - 1;

            while (j >= 0 && values[j] > value)
            {
                values[j + 1] = values[j];
                j = j - 1;
            }
            values[j + 1] = value;
        }
    }

    //SelectionSort
    public void selectionSort(int[] values)
    {
        if (values.length <= 1)
            return;

        int j, minIndex;
        for (int i = 0; i < values.length - 1; i++)
        {
            minIndex = i;
            for (j = i + 1; j < values.length; j++)
            {
                if (values[j] < values[minIndex])
                {
                    minIndex = j;
                }
            }

            int temp = values[minIndex];
            values[minIndex] = values[i];
            values[i] = temp;
        }
    }

    //MergeSort
    public void mergeSort(int[] values, int left, int right)
    {
        if (left == right)
            return;

        if (left < right) {
            int middle = (left + right) / 2;

            mergeSort(values, left, middle);
            mergeSort(values, middle + 1, right);

            int[] temp = new int[values.length];
            for (int n = left; n <= right; n++) {
                temp[n] = values[n];
            }

            int index1 = left;
            int index2 = middle + 1;
            for (int n = left; n <= right; n++) {
                if (index1 == middle + 1) {
                    values[n] = temp[index2++];
                } else if (index2 > right) {
                    values[n] = temp[index1++];
                } else if (temp[index1] < temp[index2]) {
                    values[n] = temp[index1++];
                } else {
                    values[n] = temp[index2++];
                }
            }
        }
    }

    //QuickSort
    public void quickSort(int[] values, int low, int high)
    {
        if (low < high)
        {
            int index = partition(values, low, high);

            quickSort(values, low, index - 1);
            quickSort(values, index + 1, high);
        }
    }

    int partition(int[] values, int low, int high)
    {
        int pivot = values[high];
        int i = (low - 1);
        for (int j = low; j <= high - 1; j++)
        {
            if (values[j] <= pivot)
            {
                i++;

                int temp = values[i];
                values[i] = values[j];
                values[j] = temp;
            }
        }

        i++;
        int temp = values[i];
        values[i] = values[high];
        values[high] = temp;

        return i;
    }

    //BubbleSort
    public void bubbleSort(int[] values)
    {
        boolean swapped;
        for (int i = 0; i < values.length - 1; i++)
        {
            swapped = false;
            for (int j = values.length -1; j > i; j--)
            {
                if (values[j] < values[j - 1])
                {
                    int temp = values[j];
                    values[j] = values[j - 1];
                    values[j - 1] = temp;
                    swapped = true;
                }
            }

            if (swapped == false)
                break;
        }
    }

    //BucketSort
    public void BucketSort(int[] values, int maxVal)
    {
        int[] bucket = new int[maxVal + 1];
        int num = values.length;
        int bucketNum = bucket.length;

        for (int i = 0; i < bucketNum; i++)
        {
            bucket[i] = 0;
        }

        for (int i = 0; i < num; i++)
        {
            bucket[values[i]]++;
        }

        int pos = 0;
        for (int i = 0; i < bucketNum; i++)
        {
            for (int j = 0; j < bucket[i]; j++)
            {
                values[pos++] = i;
            }
        }
    }

    //CountingSort
    public void CountingSort(int[] values)
    {
        int n = values.length;
        int min = values[0];
        int max = values[0];
        for (int i = 1; i < n; i++)
        {
            if (values[i] < min)
            {
                min = values[i];
            }
            else if (values[i] > max)
            {
                max = values[i];
            }
        }

        int[] count = new int[max - min + 1];
        int z = 0;

        for (int i = 0; i < n; i++)
        {
            count[i] = 0;
        }

        for (int i = 0; i < n; i++)
        {
            count[values[i] - min]++;
        }

        for (int i = min; i <= max; i++)
        {
            while (count[i - min]-- > 0)
            {
                values[z] = i;
                ++z;
            }
        }
    }

    //RadixSort
    public void RadixSort(int[] values)
    {
        int n = values.length;
        int max = values[0];
        for (int i = 1; i < n; i++)
        {
            if (values[i] > max)
            {
                max = values[i];
            }
        }

        for (int exp = 1; max / exp > 0; exp *= 10)
        {
            int[] output = new int[n];
            int[] count = new int[10];

            // Store count of occurrences in count[]
            for (int i = 0; i < n; i++)
            {
                count[(values[i] / exp) % 10]++;
            }

            for (int i = 1; i < 10; i++)
            {
                count[i] += count[i - 1];
            }

            for (int i = n - 1; i >= 0; i--)
            {
                output[count[(values[i] / exp) % 10] - 1] = values[i];
                count[(values[i] / exp) % 10]--;
            }

            for (int i = 0; i < n; i++)
            {
                values[i] = output[i];
            }
        }
    }


}
